### PR TITLE
vlc-video: Allow media options

### PR DESCRIPTION
**Pull request by EstevanTH**
Hi, this is my first time bringing to OBS Studio!

I have added support for media options. They are important or even mandatory for some usages. Options have to begin with ':' and a space must precede the heading colon. This work is to be reviewed.

I am not really experienced in working in C with computers and I'm still disturbed while browsing source code. Although this modification is pretty important to me so I spent some time to figure out myself the lack of support for VLC media options.

Please proofread and give me tips to improve this work and my future works.

It was hard to follow the coding rules, especially the max 80 columns.

Momo :heart:

### CODE REVIEW

File: plugins/vlc-video/vlc-video-source.c [hunk 1]
```diff
@@ -394,12 +394,97 @@ static int vlcs_audio_setup(void **p_data, char *format, unsigned *rate,
 	return 0;
 }
 
+static inline void vlc_split_media_path(const struct dstr *new_path,
+	struct dstr *path_only, struct dstr *options)
+{
+	int i;
+	const char* path = new_path->array;
```
Comment (RytoEX) [inline plugins/vlc-video/vlc-video-source.c]: Style:  The asterisk should be next to the variable name when declaring a pointer.  Use "const char *path".

File: plugins/vlc-video/vlc-video-source.c [hunk 2]
```diff
@@ -394,12 +394,97 @@ static int vlcs_audio_setup(void **p_data, char *format, unsigned *rate,
...
+	bool quoting = false;
+	int options_at = 0;
+	int path_len;
+
+	for (i = 0; path[i]; ++i) {
+		if (path[i] == '"') {
+			quoting = !quoting;
+		}
+		if (!quoting && path[i+1]==' ' && path[i+2]==':') {
```
Comment (RytoEX) [inline plugins/vlc-video/vlc-video-source.c]: Style: Place spaces around the equality operator (==).  There is some inconsistency in the existing codebase about whether or not there should be spaces around the plus signs (+) inside the square brackets for the array index indicators, so I'll let @jp9000 comment on that.

File: plugins/vlc-video/vlc-video-source.c [hunk 3]
```diff
@@ -394,12 +394,97 @@ static int vlcs_audio_setup(void **p_data, char *format, unsigned *rate,
...
+			// ++i to be consistent with finished loop case
+			options_at = ++i;
+			break;
+		}
+	}
+	if (options_at) { // grab options
+		dstr_copy(options, new_path->array+options_at);
```
Comment (RytoEX) [inline plugins/vlc-video/vlc-video-source.c]: Style: Places spaces around the plus sign (+).

File: plugins/vlc-video/vlc-video-source.c [hunk 4]
```diff
@@ -394,12 +394,97 @@ static int vlcs_audio_setup(void **p_data, char *format, unsigned *rate,
...
+	}
+	for(--i; path[i]==' '; --i); // remove trailing spaces
```
Comment (RytoEX) [inline plugins/vlc-video/vlc-video-source.c]: Style: Place spaces around the equality operator (==).

File: plugins/vlc-video/vlc-video-source.c [hunk 5]
```diff
@@ -394,12 +394,97 @@ static int vlcs_audio_setup(void **p_data, char *format, unsigned *rate,
...
+	path_len = i+1;
```
Comment (RytoEX) [inline plugins/vlc-video/vlc-video-source.c]: Style: Places spaces around the plus sign (+).

File: plugins/vlc-video/vlc-video-source.c [hunk 6]
```diff
@@ -394,12 +394,97 @@ static int vlcs_audio_setup(void **p_data, char *format, unsigned *rate,
...
+	dstr_ncopy_dstr(path_only, new_path, path_len);
+}
+
+static inline void vlc_apply_media_options(libvlc_media_t *media,
+	const char *options, const bool is_url)
```
Comment (RytoEX) [inline plugins/vlc-video/vlc-video-source.c]: Style: Use two tabs for a continued/multi-line function declaration.

File: plugins/vlc-video/vlc-video-source.c [hunk 7]
```diff
@@ -394,12 +394,97 @@ static int vlcs_audio_setup(void **p_data, char *format, unsigned *rate,
...
+{
+	int i, j;
```
Comment (RytoEX) [inline plugins/vlc-video/vlc-video-source.c]: Style: Don't make multiple variable declarations on one line.  Split this into:
```
int i;
int j;
```

File: plugins/vlc-video/vlc-video-source.c [hunk 8]
```diff
@@ -394,12 +394,97 @@ static int vlcs_audio_setup(void **p_data, char *format, unsigned *rate,
...
+	bool quoting = false;
+	int option_at = 0;
+	int option_len;
+	char* option_str = 0;
```
Comment (RytoEX) [inline plugins/vlc-video/vlc-video-source.c]: Style: The asterisk should be next to the variable name when declaring a pointer. Use "char *option_str".

File: plugins/vlc-video/vlc-video-source.c [hunk 9]
```diff
@@ -394,12 +394,97 @@ static int vlcs_audio_setup(void **p_data, char *format, unsigned *rate,
...
+		if (options[i] == '"') {
+			quoting = !quoting;
+		}
+		if (!option_at) {
+			if (options[i] == ':') {
+				// option begins
+				option_at = i;
+			}
+		}
+		else if (options[i+1] == '\0' || (!quoting &&
```
Comment (RytoEX) [inline plugins/vlc-video/vlc-video-source.c]: Style: "else if" should be on the same line as the closing curly brace above on the previous line.

File: plugins/vlc-video/vlc-video-source.c [hunk 10]
```diff
@@ -394,12 +394,97 @@ static int vlcs_audio_setup(void **p_data, char *format, unsigned *rate,
...
+			options[i+1] == ' ' && options[i+2] == ':')) {
+			// option completed
+			option_len = i+1-option_at;
```
Comment (RytoEX) [inline plugins/vlc-video/vlc-video-source.c]: Style: Places spaces around the plus sign (+) and minus sign (-).

File: plugins/vlc-video/vlc-video-source.c [hunk 11]
```diff
@@ -394,12 +394,97 @@ static int vlcs_audio_setup(void **p_data, char *format, unsigned *rate,
...
+			for (j = i; j != option_at; j--) {
+				// remove trailing spaces
+				if (options[j] == ' ') {
+					--option_len;
+				}
+				else {
```
Comment (RytoEX) [inline plugins/vlc-video/vlc-video-source.c]: Style: "else" should be on the same line as the closing curly brace on the previous line.

File: plugins/vlc-video/vlc-video-source.c [hunk 12]
```diff
@@ -394,12 +394,97 @@ static int vlcs_audio_setup(void **p_data, char *format, unsigned *rate,
...
+					break;
+				}
+			}
+			option_str = malloc((option_len+1) *
```
Comment (RytoEX) [inline plugins/vlc-video/vlc-video-source.c]: Style: Places spaces around the plus sign (+).

File: plugins/vlc-video/vlc-video-source.c [hunk 13]
```diff
@@ -394,12 +394,97 @@ static int vlcs_audio_setup(void **p_data, char *format, unsigned *rate,
...
+				sizeof(char));
+			memcpy(option_str, options+option_at, option_len);
```
Comment (RytoEX) [inline plugins/vlc-video/vlc-video-source.c]: Style: Places spaces around the plus sign (+)

File: plugins/vlc-video/vlc-video-source.c [hunk 14]
```diff
@@ -394,12 +394,97 @@ static int vlcs_audio_setup(void **p_data, char *format, unsigned *rate,
...
+			option_str[option_len] = '\0';
+			libvlc_media_add_option_(media, option_str);
+			if (!custom_network_caching &&
+				option_len > 17) {
+				custom_network_caching = strncmp(
+					option_str,
+					":network-caching=", 17)==0;
```
Comment (RytoEX) [inline plugins/vlc-video/vlc-video-source.c]: Style: Place spaces around the equality operator (==).

**Comment by jp9000** [comment]
Hey there, nice to meet you, and welcome.  Out of curiosity, what specific VLC options have you tested with this?

(Terribly sorry about the formatting issues by the way -- I should have just used .clang-format auto-formatting from the beginning, but using it this late in to the project would mess with the history significantly, so we're avoiding it at the moment)

**Comment by SuslikV** [comment]
Any way we could use this to workaround this one: https://obsproject.com/mantis/view.php?id=860 ?

**Comment by EstevanTH** [comment]
Hi!

I just had an idea to improve my work. I parse quotes as delimiters, but the media path given to libVLC does not support them. However, despite quoting not being allowed, the ability to quote the path seems natural to me when parameters may appear after it. So I guess I should remove quotes from the path given to libVLC and make life easier. 😛

@RytoEX: Thanks for the coding style notes, I'll fix that very soon.

@jp9000: Thanks! I have tested successfully `:network-caching=...`, `:screen-fps=...`, `:demux=...` and I have *successlessly* 😆 tested `:repeat`, which does not seem to be a media option.
I just have successfully tested the media string `dvb-t://frequency=554167000:bandwidth=8 :dvb-adapter=0 :live-caching=2000 :program=515` to grab a DVB-T channel (deinterlacing is not media options 😢). I guess that capturing DirectShow devices is totally possible as well.
In my case I absolutely needed to change RTP options, but I'm not sure yet whether it changed anything or not `rtp/ts://@192.168.**.**:**** :network-caching=5000 :rtp-max-dropout=32767 :rtp-max-misorder=32767 :no-avcodec-dr`.

@SuslikV: Well actually I have more information about that said VLC resolution issue. It seems do depend on how the video was coded. In some cases [MediaInfo](https://mediaarea.net/) is able to report the weird VLC resolution. If you do not know the size in advance, maybe there are filters that you can use, I do not really know.

**Comment by EstevanTH** [comment]
Hello, tell me what you guys think now! 😛

File: plugins/vlc-video/vlc-video-source.c [hunk 15]
```diff
@@ -394,12 +394,115 @@ static int vlcs_audio_setup(void **p_data, char *format, unsigned *rate,
...
+			option_len = i + 1 - option_at;
+			for (j = i; j != option_at; j--) {
+				// remove trailing spaces
+				if (options[j] == ' ') {
+					--option_len;
+				} else {
+					break;
+				}
+			}
+			option_str = malloc((option_len + 1) *
```
Comment (RytoEX) [inline plugins/vlc-video/vlc-video-source.c]: Style:  Lines 460-461 can actually be one line since they do not exceed 80 characters.

Instead of:
```
			option_str = malloc((option_len + 1) *
				sizeof(char));
```

use:
```
			option_str = malloc((option_len + 1) * sizeof(char));
```

File: plugins/vlc-video/vlc-video-source.c [hunk 16]
```diff
@@ -394,12 +394,115 @@ static int vlcs_audio_setup(void **p_data, char *format, unsigned *rate,
...
+					++k;
+					++j;
+				} else {
+					option_str[k] = option_str[j];
+					++k;
+				}
+			}
+			option_str[k] = '\0';
+			libvlc_media_add_option_(media, option_str);
+			if (!custom_network_caching &&
```
Comment (RytoEX) [inline plugins/vlc-video/vlc-video-source.c]: Style: Lines 482-483 can actually be one line since they do not exceed 80 characters.

Instead of:
```
			if (!custom_network_caching &&
				option_len > 17) {
```

use:
```
			if (!custom_network_caching && option_len > 17) {
```

**Comment by EstevanTH** [comment]
Thank you, you have a good eye! 👀
Ugh, if I have to squash then the begin of my lost state is beginning!

**Comment by EstevanTH** [comment]
Okay it seems that the job is all done! I am waiting for further instructions to make it ready to be merged. 😋

For my next pull request I will probably try to add the possibility to use VLC general options (passed to `libvlc_new()`), even though there is a warning (more applicable to permanent options):
> There is absolutely no warranty or promise of forward, backward and cross-platform compatibility with regards to libvlc_new() arguments. We recommend that you do not use them, other than when debugging.

**Comment by EstevanTH** [comment]
Unfortunately it's not as easy with global VLC options... 😢
[vlc-video: Allow global options {GitHub}](https://github.com/EstevanTH/obs-studio/commit/7843d66f0018e1c170604aa03c8969d7fbc363cd)
[vlc-video: random bug while passing argv[] (threaded problem?) {forum}](https://obsproject.com/forum/threads/vlc-video-random-bug-while-passing-argv-threaded-problem.69607/)

**Comment by jp9000** [comment]
Hey there -- I chose to merge the network cache PR instead of this one.  This one is kind of complex and difficult to use unfortunately.  That and I'd prefer the individual VLC options be actual properties if there are any that should be added.

Don't let that discourage you though, it's mostly just a particular design decision on my part.

Terribly sorry about the delay, I've been putting off lower priority PRs because things have been busy these past few months.

**Comment by EstevanTH** [comment]
Okay then I'll publish my own because I need media options and I believe others also do, for RTP, DVB and many others.

I thought I really did my best to make it simple and super easy. 😞 At least it's not some lost work. 😛

**Comment by jp9000** [comment]
What specific media options?

**Comment by EstevanTH** [comment]
I have edited my message weeks ago, I gave all options that I was using. I put bold where I need.
- `:network-caching=...`: useless with the new option
- `:screen-fps=...,` to replace screen capture fps (useless because who uses VLC screen capture)
- `:demux=...` to replace demuxer (mostly useless)
- *`:dvb-adapter=0`*: I do not know if it is mandatory when having just 1 tuner
- *`:live-caching=2000`*: maybe default value is alright anyway
- *`:program=515`*: mandatory to select a TV/radio channel on with a tuner
- *`:rtp-max-dropout=32767`*: the default value does packet losses because of a huge packet misorder in my case
- *`:rtp-max-misorder=32767`*: same as the previous
- `:no-avcodec-dr`: no idea if it has something to do with *libVLC*, but *VLC Media Player* was not happy and constantly re-enables & disables direct rendering when I was playing my RTP stream (although it has probably nothing to do with RTP playback), so I used it

It's a matter of freedom. My attempt here was to bring freedom to other users.

**Comment by EstevanTH** [comment]
Now that my bad feelings are 70% gone, I can provide a thought answer.

I do not understand why this PR has been denied, simply because it's totally transparent to the user. If the user does not use any option, it's alright, there is no change anywhere in this case. If the user needs options, he just enters them after the path / MRL and here we go. I added support for quotes in case of problems with spaces, but still it's transparent to anyone not using options.

Okay I have to admit something, if the path is not quoted and contains the exact sequence ` :` then it will be understood as the beginning of an option. As this is not likely to happen and there is the quoting solution, I think that it's totally worth accepting this tiny problem.

I do not understand how this is complicated either. It's just about copying and pasting advanced options displayed in the Open box in *VLC Media Player*.

For anybody interested, I have published [my own `vlc-video` plugin featuring VLC deinterlace and media options](https://github.com/EstevanTH/OBS_Studio_modded_plugins/releases/tag/2017-06-26), based on the latest version.

**Comment by jp9000** [comment]
Your feelings are understandable, and you may think that there's no reason not to just because it doesn't hurt anyone to have the feature you want in, but my reasoning has nothing to do with what it provided, what it provided was fine, but instead is almost entirely UX (user experience) related.

1.) It's a niche feature that is only for very specific and pretty rare cases,

2.) It's difficult to document, the formatting required is weird, sort of like URL parameters but not really if that makes sense, they'd have to make this long string at the end of their URL starting with the ":" character.  They'd have to browse the documentation to even understand how to do it.  If they mess up or make a mistake by accident with it, they may not realize they've done so, or you'd have to implement some convoluted way of communicating that to the user,

3.) When the user is using the program, they have no idea it's even there, it's impossible for the user to know that this feature is available to them, again they'd have to browse the documentation to even know it's there, where as the network caching feature however is visible to every user who wants to use it and it's easy to use and doesn't require making a string at the end of your URL to be able to use it.

That is why I said if we want additional options, they should be explicit properties that are easy to use, that every user knows is there, and not some string that you have to mangle to the end of the URL.  I hope you understand my reasoning now behind why I decided to respectfully decline this particular patch.

----

In addition to the above reasons, I realize that you might say "it hurts no one for it to be there", but that's not quite true:  any code that's added to the repository adds a maintenance cost, however small, and that's something that people don't often realize when they say "my code doesn't hurt anybody being there", it hurts the people who have to take care of the project.  When a contributor comes along and says "hey pull my code", 99% of the time the maintainers of the project are the ones who have to maintain that code if there's a problem with it (directly or indirectly) in the future.  Even if it's just a small patch, it piles more code in to the project and decreases the quality of life for the maintainer because the repository just gets endlessly messy with these super niche, often almost-never-used features.

So your feelings are understandable, but there are very good reasons that I think you might not quite have realized.  Your code itself works, and its functionality is perfectly fine, that's not why I'm declining your code.

I hope you understand, and I hope it does not discourage you from continuing contribution, but instead is something to learn from so your future contributions take these factors in to account.  I understand that having to deal with maintainers isn't always fun/fair, but I hope that even despite declining, that I have been as polite/respectful as possible.

**Comment by EstevanTH** [comment]
Yes you have been very respectful and explicit, thanks for that.

I know that more code is more maintenance time. This said, I'm still at the step of programming experience when I think that the number of useful possibilities in using a library in a given program is a finished number. And if I had to imagine the whole possibilities, allowing media options is for me a big part amongst the others.

Now I have said what I wanted to tell. See you next post!

**Comment by kingargyle** [comment]
@EstevanTH  I'm not one of the developers, but one thing that might help from a UI and user stand point to address one of @jp9000  concerns, is having an Advanced option in the UI for the VLC plugin itself.  Something that isn't readily enabled all the time, and clearly marks it as advanced territory.  From here, you could have a Properties box that allows the appropriate key value pair to be applied. 

The other option would be to make a UI that supports various options you wanted to add by default and gives the option to add additional ones free hand.  Again I'd put this all under an advanced configuration section of the vlc plugin's properties dialog.   It would make it a little more user friendly, and in general the user may not need to know what the exact format should be.  Just a suggestion.

**Comment by EstevanTH** [comment]
Hello. I believe that this is a nice idea. Then my question to @jp9000 is: can we modify the interface used to load modules? I think it's not so much of a good idea, but maybe. It's like a switch to enable "experimental" things.

If I put a boolean to disabled advanced stuff, then maybe it's safer to add such features that are somehow dangerous because they are like a knife in a string. 😃

**Comment by ch-sirius** [comment]
Hi there. Thank you for your valuable contributions. I'm personally waiting for that feature since weeks. We plan to use OBS for a live stream with 12 network cameras. There we need to optimize some stream properties to get a good video quality. So, the possibility of adding vlc options is crucial in this project. @EstevanTH could you please upload a build of the plugin with your latest version for 64bit, too? Thank you!

@jp9000 of course I understand your points. But I think if you use VLC the [MRL Syntax](https://wiki.videolan.org/Media_resource_locator/) should be supported. People who need this options know how to use them (the same way as in vlc).

**Comment by EstevanTH** [comment]
Hi @ch-sirius I'm glad you appreciate it. Building a Windows AMD64 version should not be too difficult so I'm going to do it and I come back to you when it's done.

**Comment by EstevanTH** [comment]
It's done, but of course untested because I cannot have both VLC 32-bit and 64-bit on the same machine!

May you find it now: [VLC source (Windows IA-32 and AMD64) 2017-06-26](https://github.com/EstevanTH/OBS_Studio_modded_plugins/releases/tag/2017-06-26)

**Comment by ch-sirius** [comment]
Thank you @EstevanTH! It works like a charm.
